### PR TITLE
App versioning

### PR DESCRIPTION
**Pull request by sajithar**
Implementing feature: Improving app versioning support in App Manager

### CODE REVIEW

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 1]
```diff
@@ -21,553 +21,805 @@
...
+		});
+
+		for (var i = 0; i < criticalRoles.length; i++) {
+			if (allRoles.indexOf(criticalRoles[i]) > -1) {
+				allRoles.splice(i, 1);
+			}
+		}
+		;
+
+		// TODO: needs to go in a config
```
Comment (jenananthan) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Fix the to do , if already fixed remove the todo comment

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml [hunk 2]
```diff
@@ -0,0 +1,110 @@
+<!--
+ ~ Copyright (c) 2005-2015, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lahiru823) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: Check copyright format

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 3]
```diff
@@ -21,553 +21,805 @@
...
+		var carbon = require('carbon');
+		var realmService = carbon.server.osgiService('org.wso2.carbon.user.core.service.RealmService');
+		var realm = realmService.getTenantUserRealm(server.current(session).tenantId);
+
+		var allRoles = realm.getUserStoreManager().getRoleNames();
+
+		var lifecycleJSON = obtainLifeCycle(shortName);
+		var relatedRoles = [];
+
+		// TODO: needs to go in a config
```
Comment (jenananthan) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: fix the todo, if already fixed remove the comment

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 4]
```diff
@@ -21,553 +21,805 @@
...
+			if (invokeAction != '') {
+				artifactManager.promoteLifecycleState(invokeAction, artifact);
+			}
+
+			var msg = {};
+			msg['status'] = 'ok';
+
+			print(msg);
+		} catch (e) {
+			log.debug('The asset with id: ' + id + ' could not be found.The following exception was thrown: ' + e);
```
Comment (lahiru823) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Shouldn't this be traced in log.error() level?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 5]
```diff
@@ -21,553 +21,805 @@
...
+		var artifactManager = rxtManager.getArtifactManager(shortName);
+
+		try {
+			var artifact = artifactManager.get(id);
+			artifactManager.checkItem(checkListItemIndex, artifact);
+			var msg = {};
+			msg['status'] = 'ok';
+		} catch (e) {
+			log.debug('The asset with id: ' + id + ' could not be found.The following exception was thrown: ' + e);
+			response.sendError(404, 'The asset with id: ' + id + ' could not be found.Please refer to the server log.');
```
Comment (lahiru823) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Shouldn't this be traced in log.error() level?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 6]
```diff
@@ -21,553 +21,805 @@
...
+		} else {
+			log.debug('Does not have permission to perform the action');
+			msg['status'] = 'failed';
+			msg['message'] = 'You do not have rights to perform the life-cycle operation';
+			print(msg);
+			response.status = 401;
+		}
+	}
+	catch (e) {
+		log.debug('The asset with id: ' + id + ' could not be found.The following exception was thrown: ' + e);
```
Comment (lahiru823) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Shouldn't this be traced in log.error() level?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 7]
```diff
@@ -21,553 +21,805 @@
...
+				var lcHistoryXML = new XML(lcHistory.content);
+
+				lcHistoryJSON = utility.xml.convertE4XtoJSON(lcHistoryXML);
+				var newChangeOrder = lcHistoryJSON.item[0].order;
+				require('/modules/comment.js').addLifecycleComment(artifactManager, artifact.path, comment, newChangeOrder);
+			}
+			msg['status'] = 'ok';
+			print(msg);
+		} else {
+			log.debug('Does not have permission to perform the action');
```
Comment (lahiru823) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: check if isDebugEnabled()

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+					<property name="Host" scope="transport" action="remove"/>
+					<property name="Accept" scope="transport" action="remove"/>
+					<send/>
+				</then>
+				<else>
+					<header name="WSO2_APPM_INVOKED_WITHOUT_VERSION" scope="transport" value="true"/>
+					<!--synapse variables in 'uri-template' tags MUST begin with 'uri.var.'-->
+					<property name="uri.var.ip" expression="get-property('synapse.host.ip')"/>
+					#if( $transport == "https" )
+					<property name="uri.var.portnum" expression="get-property('https.nio.port')"/>
```
Comment (sajithab) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: Formatting? L71 as well
Isn't it possible to use <property name="uri.var.portnum" as <property name="uri.var.portNumber"? L71 as well

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+					#else
+					<property name="uri.var.portnum" expression="get-property('http.nio.port')"/>
+					#end
+					<send>
+						<endpoint>
+							## change localhost to ip address
+							#if( $transport == "https" )
+							<http uri-template="https://{uri.var.ip}:{uri.var.portnum}/$!{forwardAppContext}/${forwardAppVersion}">
```
Comment (sajithab) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: Formatting? L79 as well.

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    #end
+
+    <send>
+        #if ($endpointClass == "http")
+            #http_endpoint($ep_name $endpoints)
+        #elseif ($endpointClass == "address")
+            <endpoint xmlns="http://ws.apache.org/ns/synapse" name="$ep_name">
+                #set( $advance_ep = $endpoints.get("config") )
+                <address uri="$util.escapeXml($endpoints.get("url"))"
+                    #if($advance_ep.get("format") && ($advance_ep.get("format") != "") && ($advance_ep.get("format") != 'leave-as-is'))
```
Comment (sajithab) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml]: Is this allowed in synapse? Checking string with single quote? ($advance_ep.get("format") != 'leave-as-is'

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 11]
```diff
@@ -21,553 +21,805 @@
...
+	var router = require('/modules/router-g.js').router();
+	var routeManager = new router.Router();
+	var log = new Log();
+	var permissions = require('/modules/permissions.js').permissions;
+
+	routeManager.register('POST', 'publisher', '/publisher/api/lifecycle/{type}/{id}', function (context) {
+		var shortName = context.params.type;
+		var id = context.params.id;
+		var lifecycle = context.post.lifecycle;
+		//var autoPromote=context.post.promote||false;    //If promote is not specified do not auto promote
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Remove the commented code

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 12]
```diff
@@ -21,553 +21,805 @@
...
+	var config = require('/config/publisher.json');
+	var caramel = require('caramel');
+	var router = require('/modules/router-g.js').router();
+	var routeManager = new router.Router();
+	var log = new Log();
+	var permissions = require('/modules/permissions.js').permissions;
+
+	routeManager.register('POST', 'publisher', '/publisher/api/lifecycle/{type}/{id}', function (context) {
+		var shortName = context.params.type;
+		var id = context.params.id;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: better to use meaningful name like webAppId

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 13]
```diff
@@ -21,553 +21,805 @@
...
+		//var autoPromote=context.post.promote||false;    //If promote is not specified do not auto promote
+		var invokeAction = '';
+		//Check the config for a lifeCycleBehaviour block
+		utility.isPresent(config, 'lifeCycleBehaviour', function (lifeCycleBehaviour) {
+			utility.isPresent(lifeCycleBehaviour, lifeCycle, function (lifeCycleData) {
+				utility.isPresent(lifeCycleData, 'onAttach', function (onAttach) {
+					invokeAction = onAttach.action || '';
+				});
+			});
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Formatting?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 14]
```diff
@@ -21,553 +21,805 @@
...
+				artifactManager.promoteLifecycleState(invokeAction, artifact);
+			}
+
+			var msg = {};
+			msg['status'] = 'ok';
+
+			print(msg);
+		} catch (e) {
+			log.debug('The asset with id: ' + id + ' could not be found.The following exception was thrown: ' + e);
+			response.sendError(404, 'The asset with id: ' + id + ' could not be found.Please refere to the server log.');
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: End user will not be able to view server logs. Fix the error message.

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag (hunk 4 above)
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: found.The -> found. The

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml (hunk 10 above)
Comment (sajithar) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml]: It i allowed & works without any issue. This code segment was already there, I just formatted it.

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 15]
```diff
@@ -21,553 +21,805 @@
...
+		var id = context.params.id;
+		var artifactManager = rxtManager.getArtifactManager(shortName);
+
+		try {
+			var artifact = artifactManager.get(id);
+			artifactManager.uncheckItem(checkListItemIndex, artifact);
+			var msg = {};
+			msg['status'] = 'ok';
+		} catch (e) {
+			log.debug('The lifecyle check list item at index ' + checkListItemIndex + ' of asset id: ' + id + ' could not be unticked' + 'The following exception was thrown: ' + e);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: unticked' + 'The -> unticked. ' + 'The

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 16]
```diff
@@ -21,553 +21,805 @@
...
+				currentArtifact.attributes.overview_provider);
+			log.error(e);
+			// continue work flow
+		}
+
+		var operationResult; // {isSuccess: true/false, message: ""}
+
+		// for all previously published apps when publishing or re-publishing current app
+		// unpublish previously published version of this app
+		if (((action == "Publish") || (action == "Re-Publish")) && (previousArtifacts.length > 0)) {
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Better to use "Publish" and "Re-Publish" as constant. There are number of usage of those in this class.

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 17]
```diff
@@ -21,553 +21,805 @@
...
-});
-
-var securityModule=require('/modules/security/security.manager.js').securityManagementModule();
-
-var sm=securityModule.cached();
+	/*
+	 The function returns the lifecycle structure of a given asset
+	 */
+	function obtainLifeCycle(type) {
+		CommonUtil = Packages.org.wso2.carbon.governance.lcm.util.CommonUtil;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Is there any reason to put this object in global context? 
CommonUtil -> commonUtil

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 18]
```diff
@@ -21,553 +21,805 @@
...
+		//Obtain a model of the type
+		var model = modelManager.getModel(type);
 
-  //Stop servicing the request if the check failed
-  if (passed) {
-    var store = require('store');
-    var user = store.user;
-    var utility = require('/modules/utility.js').rxt_utility();
-    var server = store.server;
+		var lifecycle = model.get('*.lifeCycle').value;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: lifecycle -> lifeCycle

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 19]
```diff
@@ -21,553 +21,805 @@
...
+		data = data.replace('<?xml version="1.0" encoding="ISO-8859-1"?>', '');
+		var lcJSON = utility.xml.convertE4XtoJSON(xml);
 
+		return lcJSON;
+	}
 
+	routeManager.register('GET', 'publisher', '/publisher/api/lifecycle/{type}/{id}', function (context) {
+		try {
+
+			appPublishWFExecutor = org.wso2.carbon.appmgt.impl.workflow.WorkflowExecutorFactory.getInstance().getWorkflowExecutor("AM_APPLICATION_PUBLISH");
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Is there any reason to put this object in global context?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 20]
```diff
@@ -21,553 +21,805 @@
...
+			var isAsynchronousFlow = appPublishWFExecutor.isAsynchronus();
+			var shortName = context.params.type;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: shortName ?? is this appType?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 21]
```diff
@@ -21,553 +21,805 @@
...
+			var lcJSON = obtainLifeCycle(shortName);
+
+			var msg = {};
+			msg['status'] = 'ok';
+			msg['state'] = state;
+			msg['actions'] = actions;
+			msg['lifecycle'] = lcJSON;
+			msg['asynch'] = isAsynchronousFlow;
+
+			log.debug(msg);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Check if (log.isDebugEnabled()) {

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 22]
```diff
@@ -21,553 +21,805 @@
...
+			var subscribed = false;
+			if (count > 0) {
+				subscribed = true;
+			}
+
+			var msg = {};
+			msg['subscribed'] = subscribed;
+			msg['state'] = state;
+
+			log.debug(msg);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Check if (log.isDebugEnabled()) {

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 23]
```diff
@@ -21,553 +21,805 @@
...
+
+		for (var i = 0; i < allRoles.length; i++) {
+			if (allRoles[i].indexOf(query) > -1) {
+				matchedRoles.push({
+					id: allRoles[i],
+					name: allRoles[i]
+				});
+			}
+		}
+		;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Formatting?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 24]
```diff
@@ -21,553 +21,805 @@
...
+
+		matchedRoles = parse(stringify(matchedRoles));
+
+		var hiddenRoles = config.hiddenRoles;
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Blank new line?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 25]
```diff
@@ -21,553 +21,805 @@
...
+		if (matchedRoles.length > 0) {
+			for (var i = 0; i < matchedRoles.length; i++) {
+				if ((matchedRoles[i].id.indexOf("Internal/private_") == -1) && (hiddenRoles.indexOf(matchedRoles[i].id) == -1)) {
+					filteredRoles.push(matchedRoles[i]);
+				}
+			}
+		}
+
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Blank new line?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 26]
```diff
@@ -21,553 +21,805 @@
...
+		response.contentType = 'application/json';
+		response.status = 200;
+		response.content = filteredRoles;
+	});
+
+	routeManager.register('GET', 'publisher', '/publisher/api/lifecycle/checklist/{type}/{id}', function (context) {
+		//log.debug('short name: '+context.params.type);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Remove commented codes

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 27]
```diff
@@ -21,553 +21,805 @@
...
+
+		try {
+			log.debug('Entered check list/type/id ');
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Check if (log.isDebugEnabled()) {

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 28]
```diff
@@ -21,553 +21,805 @@
...
+			response.sendError(404, 'The check list item names could not be retrieved for artifact: ' + id + ' .Please refer to the server log.');
+		}
+	});
+
+	/*
+	 Gets the checked or unchecked state of a check list item
+	 */
+	routeManager.register('GET', 'publisher', '/publisher/api/lifecycle/checklistitem/{index}/{type}/{id}', function (context) {
+		try {
+			log.debug('Entered checklistitem/index/type/id ');
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Check if (log.isDebugEnabled()) {

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 29]
```diff
@@ -21,553 +21,805 @@
...
+			response.sendError(404, 'The asset with id: ' + id + ' could not be found.Please refere to the server log.');
+		}
+	});
+
+	routeManager.register('GET', 'publisher', '/publisher/api/lifecycle/information/meta/{type}/roles', function (context) {
+		var shortName = context.params.type;
+		var carbon = require('carbon');
+		var realmService = carbon.server.osgiService('org.wso2.carbon.user.core.service.RealmService');
+		var realm = realmService.getTenantUserRealm(server.current(session).tenantId);
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Blank new line?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 30]
```diff
@@ -21,553 +21,805 @@
...
+		var allRoles = realm.getUserStoreManager().getRoleNames();
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Blank new line?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 31]
```diff
@@ -21,553 +21,805 @@
...
+		var lifecycleJSON = obtainLifeCycle(shortName);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: lifecycleJSON -> lifeCycleJSON

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 32]
```diff
@@ -21,553 +21,805 @@
...
+				})
+			}
+		});
+
+		for (var i = 0; i < criticalRoles.length; i++) {
+			if (allRoles.indexOf(criticalRoles[i]) > -1) {
+				allRoles.splice(i, 1);
+			}
+		}
+		;
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Formatting?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 33]
```diff
@@ -21,553 +21,805 @@
...
+	/*
+	 Gets the checked or unchecked state of a check list item
+	 */
+	routeManager.register('GET', 'publisher', '/publisher/api/lifecycle/checklistitem/{index}/{type}/{id}', function (context) {
+		try {
+			log.debug('Entered checklistitem/index/type/id ');
+			var checkListItemIndex = context.params.index;
+			var shortName = context.params.type;
+			var id = context.params.id;
+
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Blank new line?

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 34]
```diff
@@ -21,553 +21,805 @@
...
-        artifactManager.promoteLifecycleState(action,artifact);
+	var shortName = context.params.type;
+	var action = context.params.action;
+	var id = context.params.id;
 
-        var comment;
-        if (context.content && (comment=parse(context.content).comment) ) {
-            var lcHistory = artifactManager.getLifecycleHistory(artifact);
-            var lcHistoryXML = new XML(lcHistory.content);
+	//var artifactManager = rxtManager.getArtifactManager(shortName);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Remove commented code

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 35]
```diff
@@ -21,553 +21,805 @@
...
-            response.status = 401;
-        }
-    }
-    catch (e) {
-        log.debug('The asset with id: '+id+' could not be found.The following exception was thrown: '+e);
-        response.sendError(404,'The asset with id: '+id+' could not be found.Please refer to the server log.');
-    }
+	try {
+		var artifact = artifactManager.get(id);
+		log.debug('Calling promoteLifecycleState for action : ' + action);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Check if (log.isDebugEnabled()) {

File: modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag [hunk 36]
```diff
@@ -21,553 +21,805 @@
...
+		var attributes = artifact.attributes;
 
+		if (permissions.isLCActionsPermitted(user.username, artifact.path, um)) {
+			var previourStatus = artifact['lifecycleState'];
+			log.info('Calling promoteLifecycleState for action mobile : ' + action);
```
Comment (sajithab) [inline modules/jaggery-apps/app-publisher-web/apis/v1/lifecycle_api_router.jag]: Is this really necessary to log as info log?

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml (hunk 2 above)
Comment (sajithar) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: fixed in commit cc7f512ebecdfaa5b7be5b4ea8b38b0bddef1475 of PR https://github.com/wso2/product-app-manager/pull/303

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml (hunk 9 above)
Comment (sajithar) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: fixed in commit d2c1ecb4e0bfc8bc7c102af577549f0390846f48 of PR #303

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml (hunk 8 above)
Comment (sajithar) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_non-versioned-webapp.xml]: fixed in commit b9a09ada9f3fec0afa11efea4378cdb73114a927 of PR #303

File: modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml [hunk 37]
```diff
@@ -0,0 +1,270 @@
+<!--
```
Comment (sajithar) [inline modules/distribution/resources/api_templates/velocity-template_synapse-config_versioned-webapp.xml]: Please note that this diff of the file 'velocity-template_synapse-config_versioned-webapp.xml'  is the result of moving, renaming, & code-formatting of the old file '[modules/distribution/resources/api_templates/velocity_template.xml](https://github.com/wso2/product-app-manager/pull/234/files#diff-20e21fea2ad1a1eb1057ee3a8d12d676L1)'.

**Pull request merged**